scipy/optimize/_qap.py

```diff
@@ -0,0 +1,784 @@
+    n = A.shape[0]  # number of vertices in graphs
+    n_seeds = partial_match.shape[0]  # number of seeds
+    n_unseed = n - n_seeds
+
+    obj_func_scalar = 1
+    if maximize:
+        obj_func_scalar = -1
+
+    nonseed_B = np.setdiff1d(range(n), partial_match[:, 1])
+    if shuffle_input:
+        nonseed_B = rng.permutation(nonseed_B)
+        # shuffle_input to avoid results from inputs that were already matched
+
+    nonseed_A = np.setdiff1d(range(n), partial_match[:, 0])
+    perm_A = np.concatenate([partial_match[:, 0],
+                             nonseed_A], axis=None).astype(int)
+    perm_B = np.concatenate([partial_match[:, 1],
+                             nonseed_B], axis=None).astype(int)
+
+    # definitions according to Seeded Graph Matching [2].
+    A11, A12, A21, A22 = _split_matrix(A[perm_A][:, perm_A], n_seeds)
+    B11, B12, B21, B22 = _split_matrix(B[perm_B][:, perm_B], n_seeds)
+
+    # setting initialization matrix
+    if isinstance(init_J, str) and init_J == 'barycenter':
+        J = np.ones((n_unseed, n_unseed)) / float(n_unseed)
+    else:
+        _check_init_input(init_J, n_unseed)
+        J = init_J
+
+    if init_weight != 1:
+        # generate a nxn matrix where each entry is a random number [0, 1]
+        # would use rand, but Generators don't have it
+        # would use random, but old mtrand.RandomStates don't have it
+        K = rng.uniform(size=(n_unseed, n_unseed))
+        # Sinkhorn balancing
+        K = _doubly_stochastic(K)
+        # initialize J, a doubly stochastic barycenter
+        P = J * init_weight + (1 - init_weight) * K
+    else:
+        P = J
+    const_sum = A21 @ B21.T + A12.T @ B12
+
+    # OPTIMIZATION WHILE LOOP BEGINS
+    for n_iter in range(maxiter):
+        # computing the gradient of f(P) = -tr(APB^tP^t)
+        grad_fp = (const_sum + A22 @ P @ B22.T + A22.T @ P @ B22)
+        # run hungarian algorithm on gradient(f(P))
+        rows, cols = linear_sum_assignment(obj_func_scalar * grad_fp)
+        Q = np.zeros((n_unseed, n_unseed))
+        Q[rows, cols] = 1  # initialize search direction matrix Q
+
+        def f(x):  # computing the original optimization function
+            xP1xQ = x * P + (1 - x) * Q
+            # Sums below are np.trace(A11.T @ B11)
+            # + np.trace(xP1xQ.T @ A21 @ B21.T)
+            # + np.trace(xP1xQ.T @ A12.T @ B12)
+            # + np.trace(A22.T @ xP1xQ @ B22 @ xP1xQ.T)
+            # This is more efficient, but can we do even better?
+            return obj_func_scalar * (
+                (A11 * B11).sum()
+                + (xP1xQ.T @ A21 * B21).sum()
+                + (xP1xQ.T @ A12.T * B12.T).sum()
+                + ((xP1xQ.T @ A22) * (B22 @ xP1xQ.T)).sum()
+            )
+
+        # computing the step size
+        alpha = minimize_scalar(f, bounds=(0, 1), method="bounded").x
+        P_i1 = alpha * P + (1 - alpha) * Q  # Update P
+        if np.linalg.norm(P - P_i1) < tol:
+            break
```

Comment:
```suggestion
            P = P_i1
            break
```
https://github.com/scipy/scipy/pull/12775#discussion_r478844080

Reply:
Alone, this will introduce an off by one error in the iteration count.